mobile_verifier/src/data_session.rs

```diff
@@ -0,0 +1,243 @@
+    pub pub_key: PublicKeyBinary,
+    pub payer: PublicKeyBinary,
+    pub upload_bytes: i64,
+    pub download_bytes: i64,
+    pub num_dcs: i64,
+    pub reward_timestamp: DateTime<Utc>,
+}
+
+#[derive(thiserror::Error, Debug)]
+#[error("data session error: {0}")]
+pub struct DataSessionError(#[from] sqlx::Error);
+
+impl SubscriberDataSession {
+    pub async fn save(self, db: &mut Transaction<'_, Postgres>) -> Result<(), DataSessionError> {
+        sqlx::query(
+            r#"
+            INSERT INTO subscriber_data_transfer_sessions (subscriber_id, upload_bytes, download_bytes, reward_timestamp)
+            VALUES ($1, $2, $3, $4)
+            "#,
+        )
+        .bind(self.subscriber_id)
+        // .bind("subscriber_x".to_string().as_bytes().to_vec())  // TODO: used for local debugging, remove when done
+        .bind(self.upload_bytes)
+        .bind(self.download_bytes)
+        .bind(self.reward_timestamp)
+        .execute(&mut *db)
+        .await?;
+        Ok(())
+    }
+    fn from_valid_data_session(
+        v: &ValidDataTransferSession,
+        subscriber_id: Vec<u8>,
+        reward_timestamp: DateTime<Utc>,
+    ) -> SubscriberDataSession {
+        Self {
+            subscriber_id,
+            upload_bytes: v.upload_bytes as i64,
+            download_bytes: v.download_bytes as i64,
+            reward_timestamp,
+        }
+    }
+}
+
+impl HotspotDataSession {
+    pub async fn save(self, db: &mut Transaction<'_, Postgres>) -> Result<(), DataSessionError> {
+        sqlx::query(
+            r#"
+            INSERT INTO hotspot_data_transfer_sessions (pub_key, payer, upload_bytes, download_bytes, num_dcs, reward_timestamp)
+            VALUES ($1, $2, $3, $4, $5, $6)
+            "#,
+        )
+        .bind(self.pub_key)
+        .bind(self.payer)
+        .bind(self.upload_bytes)
+        .bind(self.download_bytes)
+        .bind(self.num_dcs)
+        .bind(self.reward_timestamp)
+        .execute(&mut *db)
+        .await?;
+        Ok(())
+    }
+    fn from_valid_data_session(
+        v: ValidDataTransferSession,
+        reward_timestamp: DateTime<Utc>,
+    ) -> HotspotDataSession {
+        Self {
+            pub_key: v.pub_key,
+            payer: v.payer,
+            upload_bytes: v.upload_bytes as i64,
+            download_bytes: v.download_bytes as i64,
+            num_dcs: v.num_dcs as i64,
+            reward_timestamp,
+        }
+    }
+}
+
+pub async fn aggregate_hotspot_data_sessions_to_dc<'a>(
+    exec: impl sqlx::PgExecutor<'a> + Copy + 'a,
+    epoch: &'a Range<DateTime<Utc>>,
+) -> Result<HotspotMap, sqlx::Error> {
+    let stream = sqlx::query_as::<_, HotspotDataSession>(
+        r#"
+        SELECT *
+        FROM hotspot_data_transfer_sessions
+        WHERE reward_timestamp >= $1 and reward_timestamp < $2
+        "#,
+    )
+    .bind(epoch.start)
+    .bind(epoch.end)
+    .fetch(exec);
+    data_sessions_to_dc(stream).await
+}
+
+pub async fn data_sessions_to_dc<'a>(
```

Comment:
since `data_sessions_to_dc` is only callable by `aggregate_hotspot_data_sessions_to_dc` and `data_sessions_to_bytes` is only callable by `aggregate_subscriber_data_sessions_to_bytes` and they both return the same types can they just be collapsed from four functions into two and avoid the stream pinning?

if it's for the purposes of testing the underlying function without needing to query a database i would argue the function is really more testing the stream iteration than the aggregation of dcs/bytes since the operation it's performing is relatively simple

Reply:
These functions were split out to aid testing and to allow the tests to follow real world usage as close as possible. Agree, the aggregation is simple but the value is in how they allow the tests to be constructed